storage: TestStoreGossipSystemData failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/62f3b753d49758b22203c558b8682b394773f732

Parameters:
```
TAGS=
GOFLAGS=-race
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=795482&tab=buildLog

```
=== RUN   TestStoreGossipSystemData
I180725 05:46:18.400551 48498 gossip/gossip.go:363  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:36329" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I180725 05:46:18.416675 48498 storage/replica_command.go:279  [s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/SystemConfigSpan/Start [r2]
E180725 05:46:18.432138 48413 storage/queue.go:778  [replicate,s1,r1/1:/{Min-Table/System…}] range requires a replication change, but lacks a quorum of live replicas (0/1)
E180725 05:46:18.432829 48414 storage/queue.go:778  [replicate,s1,r2/1:/{Table/System…-Max}] range requires a replication change, but lacks a quorum of live replicas (0/1)
E180725 05:46:18.462170 48945 config/system.go:325  unable to determine largest reserved object ID from system config: descriptor table not found in system config of 0 values
I180725 05:46:18.474621 48879 storage/replica_proposal.go:222  [s1,r2/1:/{Table/System…-Max}] new range lease repl=(n1,s1):1 seq=2 start=0.000000123,101 epo=1 pro=0.000000123,123 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,4 pro=0.000000123,5
==================
WARNING: DATA RACE
Write at 0x00c420d93dd0 by goroutine 170:
  runtime.mapassign_fast64()
      /usr/local/go/src/runtime/hashmap_fast.go:522 +0x0
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).withCancel()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:242 +0x166
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).WithCancelOnStop()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x71
  github.com/cockroachdb/cockroach/pkg/storage.(*NodeLiveness).StartHeartbeat.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/node_liveness.go:459 +0xfa
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14a

Previous read at 0x00c420d93dd0 by goroutine 184:
  runtime.mapiterinit()
      /usr/local/go/src/runtime/hashmap.go:691 +0x0
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:464 +0x2d6
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).Stop.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:341 +0x2c4

Goroutine 170 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*NodeLiveness).StartHeartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/node_liveness.go:456 +0x27a
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).restartStoreWithoutHeartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:929 +0xc98
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).restartStore()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:939 +0x46
  github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreGossipSystemData()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_lease_test.go:222 +0xaa7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 184 (running) created at:
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).Stop()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:312 +0x8c
  github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreGossipSystemData()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_lease_test.go:232 +0xb4b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
==================
--- FAIL: TestStoreGossipSystemData (0.15s)
	testing.go:730: race detected during execution of test

```

--- a/pkg/util/stop/stopper.go
+++ b/pkg/util/stop/stopper.go
@@ -207,7 +207,13 @@ func (s *Stopper) RunWorker(ctx context.Context, f func(context.Context)) {
 func (s *Stopper) AddCloser(c Closer) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.mu.closers = append(s.mu.closers, c)
+	select {
+	case <-s.stopper:
+		// Close immediately.
+		c.Close()
+	default:
+		s.mu.closers = append(s.mu.closers, c)
+	}
 }
 
 // WithCancelOnQuiesce returns a child context which is canceled when the
@@ -217,7 +223,7 @@ func (s *Stopper) AddCloser(c Closer) {
 // Canceling this context releases resources associated with it, so code should
 // call cancel as soon as the operations running in this Context complete.
 func (s *Stopper) WithCancelOnQuiesce(ctx context.Context) (context.Context, func()) {
-	return s.withCancel(ctx, s.mu.qCancels)
+	return s.withCancel(ctx, s.mu.qCancels, s.quiescer)
 }
 
 // WithCancelOnStop returns a child context which is canceled when the
@@ -227,24 +233,31 @@ func (s *Stopper) WithCancelOnQuiesce(ctx context.Context) (context.Context, fun
 // Canceling this context releases resources associated with it, so code should
 // call cancel as soon as the operations running in this Context complete.
 func (s *Stopper) WithCancelOnStop(ctx context.Context) (context.Context, func()) {
-	return s.withCancel(ctx, s.mu.sCancels)
+	return s.withCancel(ctx, s.mu.sCancels, s.stopper)
 }
 
 func (s *Stopper) withCancel(
-	ctx context.Context, cancels map[int]func(),
+	ctx context.Context, cancels map[int]func(), cancelCh chan struct{},
 ) (context.Context, func()) {
 	var cancel func()
 	ctx, cancel = context.WithCancel(ctx)
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	id := s.mu.idAlloc
-	s.mu.idAlloc++
-	cancels[id] = cancel
-	return ctx, func() {
+	select {
+	case <-cancelCh:
+		// Cancel immediately.
 		cancel()
-		s.mu.Lock()
-		defer s.mu.Unlock()
-		delete(cancels, id)
+		return ctx, func() {}
+	default:
+		id := s.mu.idAlloc
+		s.mu.idAlloc++
+		cancels[id] = cancel
+		return ctx, func() {
+			cancel()
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			delete(cancels, id)
+		}
 	}
 }
 

--- a/pkg/util/stop/stopper.go
+++ b/pkg/util/stop/stopper.go
@@ -461,10 +461,13 @@ func (s *Stopper) Stop(ctx context.Context) {
 	}
 
 	s.Quiesce(ctx)
+	s.mu.Lock()
 	for _, cancel := range s.mu.sCancels {
 		cancel()
 	}
 	close(s.stopper)
+	s.mu.Unlock()
+
 	s.stop.Wait()
 	s.mu.Lock()
 	defer s.mu.Unlock()

--- a/pkg/util/stop/stopper_test.go
+++ b/pkg/util/stop/stopper_test.go
@@ -17,6 +17,7 @@ package stop_test
 import (
 	"context"
 	"fmt"
+	"runtime"
 	"sync"
 	"sync/atomic"
 	"testing"
@@ -253,6 +254,35 @@ func TestStopperClosers(t *testing.T) {
 	}
 }
 
+func TestStopperCloserConcurrent(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	const trials = 10
+	for i := 0; i < trials; i++ {
+		s := stop.NewStopper()
+		var tc1 testCloser
+
+		// Add Closer and Stop concurrently. There should be
+		// no circumstance where the Closer is not called.
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			runtime.Gosched()
+			s.AddCloser(&tc1)
+		}()
+		go func() {
+			defer wg.Done()
+			runtime.Gosched()
+			s.Stop(context.Background())
+		}()
+		wg.Wait()
+
+		if !tc1 {
+			t.Errorf("expected true; got %t", tc1)
+		}
+	}
+}
+
 func TestStopperNumTasks(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	s := stop.NewStopper()
@@ -394,6 +424,44 @@ func TestStopperWithCancel(t *testing.T) {
 	s.Stop(ctx)
 	if err := ctx2.Err(); err != context.Canceled {
 		t.Fatalf("should be canceled: %v", err)
+	}
+}
+
+func TestStopperWithCancelConcurrent(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	const trials = 10
+	for i := 0; i < trials; i++ {
+		s := stop.NewStopper()
+		ctx := context.Background()
+		var ctx1, ctx2 context.Context
+
+		// Tie two contexts to the Stopper and Stop concurrently. There should
+		// be no circumstance where either Context is not canceled.
+		var wg sync.WaitGroup
+		wg.Add(3)
+		go func() {
+			defer wg.Done()
+			runtime.Gosched()
+			ctx1, _ = s.WithCancelOnQuiesce(ctx)
+		}()
+		go func() {
+			defer wg.Done()
+			runtime.Gosched()
+			ctx2, _ = s.WithCancelOnStop(ctx)
+		}()
+		go func() {
+			defer wg.Done()
+			runtime.Gosched()
+			s.Stop(ctx)
+		}()
+		wg.Wait()
+
+		if err := ctx1.Err(); err != context.Canceled {
+			t.Errorf("should be canceled: %v", err)
+		}
+		if err := ctx2.Err(); err != context.Canceled {
+			t.Errorf("should be canceled: %v", err)
+		}
 	}
 }
 

